src/java_bytecode/expr2java.h

```diff
@@ -50,4 +52,41 @@ class expr2javat:public expr2ct
 std::string expr2java(const exprt &expr, const namespacet &ns);
 std::string type2java(const typet &type, const namespacet &ns);
 
+/// Convert floating point number to a string without printing
+/// unnecessary zeros. Prints decimal if precision is not lost.
+/// Prints hexadecimal otherwise, and/or java-friendly NaN and Infinity
+template <typename float_type>
+std::string floating_point_to_java_string(float_type value)
+{
+  static_assert(sizeof(float) == 4, "Float should be 32-bit type");
+  static_assert(sizeof(double) == 8, "Double should be 64-bit type");
```

Comment:
Why are these assertions relevant?

Reply:
Because people were complaining that in some fictitious system float is not 32 bit and double is not 64 bit.

Reply:
I don't know of any such system, but let's just assume it existed: why would it matter? That is, which piece of code relies on `float` being a 32-bit type.

Reply:
I guess correspondence between C++ float/double and Java float/double. Annoyingly, C99 nails them down to match IEEE 754, but C++11 does not.

Reply:
Why do we need to have a correspondence between Java and C++ float/double? We are generating strings, after all?

Reply:
My reading of the above code (which may be wrong) was that the `float_type` is being determined by the compiler based on the return of `ieee_floatt::to_float` which returns a `float`. If float was 64 bit, then is_float would be true, but the value would actually correspond to a Java double.

Reply:
`float_type` will be `float` if, and only if, `ieee_repr.is_double()` returns `false`. That's completely independent of the machine that this code is being compiled on.

Reply:
OK I'm convinced.